src/xdp-utils.c

```diff
@@ -117,6 +121,198 @@ needs_quoting (const char *arg)
   return FALSE;
 }
 
+static void
+xdp_usb_rule_free (XdpUsbRule *rule)
+{
+  if (!rule)
+    return;
+
+  g_free (rule);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (XdpUsbRule, xdp_usb_rule_free);
+
+static gboolean
+validate_hex_uint16 (const char *value,
+                     size_t      expected_length,
+                     uint16_t   *out_value)
+{
+  size_t len;
+  char *end;
+  long n;
+
+  g_assert (value != NULL);
+  g_assert (expected_length > 0 && expected_length <= 4);
+
+  len = strlen (value);
+  if (len != expected_length)
+    return FALSE;
+
+  n = strtol (value, &end, 16);
+
+  if (end - value != len)
```

Comment:
Not sure if this is UB or not, given that those are pointers you're doing arithmetic with.

Reply:
https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Pointer-Arithmetic.html

> C defines pointer subtraction to be consistent with pointer-integer addition, so that (p3 - p1) + p1 equals p3, as in ordinary algebra.

we should be good.